nixos/modules/services/security/kanidm.nix

```diff
@@ -189,20 +198,43 @@ in
         }
       ];
 
-    environment.systemPackages = [ pkgs.kanidm ];
+    environment.systemPackages = lib.mkIf cfg.enableClient [ pkgs.kanidm ];
 
-    systemd.services.kanidm = lib.mkIf cfg.enable {
+    systemd.services.kanidm = lib.mkIf cfg.enableServer {
       description = "kanidm identity management daemon";
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
       serviceConfig = defaultServiceConfig // {
         StateDirectory = "kanidm";
         StateDirectoryMode = "0700";
         ExecStart = "${pkgs.kanidm}/bin/kanidmd server -c ${serverConfigFile}";
-        ExecPreStart = ''
+        # There is currently no nice way to set the domain name, because origin
+        # in the config needs to be consistent with the current domain name:
+        # https://github.com/kanidm/kanidm/issues/623
+        ExecStartPre = pkgs.writeShellScript "kanidm-start-pre.sh" ''
+          set -euo pipefail
+          umask 077
+          PATH=${lib.makeBinPath (with pkgs; [ coreutils gawk gnugrep gnused ])}
+          log="$(mktemp)"
+          # Try to set domain name
           ${pkgs.kanidm}/bin/kanidmd domain_name_change \
             -c ${serverConfigFile} \
-            -n ${cfg.ensureDomainName}
+            -n ${cfg.ensureDomainName} | tee "$log"
+          # Exit status is always 0, so search for error messages
+          if grep -q 'admin.error' "$log"; then
+            # If it failed, extract the old domain name and create a temporary config
+            olddomain="$(awk 'match($0, /rp_id: "(.*)"/, m) { print m[1] }' "$log")"
+            rm "$log"
+
+            echo "Using temporary config with old domain name '$olddomain'"
+            # Create temporary config with old domain name
+            config="$(mktemp)"
+            sed ${dummyServerConfigFile} -e "s#@origin@#https://$olddomain#" > "$config"
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              -c "$config" \
+              -n ${cfg.ensureDomainName}
+            rm "$config"
+          fi
```

Comment:
Please present that hack upstream so it gets proper recognition. :facepalm:

Reply:
I'm really hesitant about that amount of shell magic in init files. Would really prefer that logic to be handled by upstream. But as it is not user facing, I'd be okay with it if it works.

Reply:
I can see what you mean :D I mentioned it in https://github.com/kanidm/kanidm/issues/623